test/stdlib/StringIndex.swift

```diff
@@ -232,21 +227,238 @@ StringIndexTests.test("String.Index(_:within) / Range<String.Index>(_:in:)") {
+      let strRange = Range<String.Index>(nsRange, in: str)
+      let substrRange = Range<String.Index>(nsRange, in: substr)
 
-        expectEqual(strRange, substrRange)
-        guard strLB != nil && strUB != nil else {
-          expectNil(strRange)
-          continue
-        }
-        expectEqual(strRange, Range(uncheckedBounds: (strLB!, strUB!)))
+      expectEqual(strRange, substrRange)
+      guard strLB != nil && strUB != nil else {
+        expectNil(strRange)
+        continue
+      }
+      expectEqual(strRange, Range(uncheckedBounds: (strLB!, strUB!)))
+    }
+  }
+}
+
+StringIndexTests.test("Misaligned") {
+  func doIt(_ str: String) {
+    let characterIndices = Array(str.indices)
+    let scalarIndices = Array(str.unicodeScalars.indices) + [str.endIndex]
+    let utf8Indices = Array(str.utf8.indices)
+
+    var lastScalarI = 0
+    for i in 1..<utf8Indices.count {
+      let idx = utf8Indices[i]
+
+      // Skip aligned indices
+      guard idx < scalarIndices[lastScalarI + 1] else {
+        assert(idx == scalarIndices[lastScalarI + 1])
+        lastScalarI += 1
+        continue
       }
+      expectTrue(UTF8.isContinuation(str.utf8[idx]))
+
+      let lastScalarIdx = scalarIndices[lastScalarI]
+
+      // Check aligning-down
+      expectEqual(str[lastScalarIdx], str[idx])
+      expectEqual(str.utf16[lastScalarIdx], str.utf16[idx])
+      expectEqual(str.unicodeScalars[lastScalarIdx], str.unicodeScalars[idx])
+
+      // Check distance
+      let (start, end) = (str.startIndex, str.endIndex)
+      if characterIndices.contains(lastScalarIdx) {
+        expectEqual(0, str.distance(from: lastScalarIdx, to: idx))
+        expectEqual(str[..<idx].count, str.distance(from: start, to: idx))
+        expectEqual(str[idx...].count, str.distance(from: idx, to: end))
+      }
+      expectEqual(
+        0, str.unicodeScalars.distance(from: lastScalarIdx, to: idx))
+      expectEqual(
+        str.unicodeScalars[..<idx].count,
+        str.unicodeScalars.distance(from: start, to: idx))
+      expectEqual(
+        str.unicodeScalars[idx...].count,
+        str.unicodeScalars.distance(from: idx, to: end))
+
+      expectEqual(0, str.utf16.distance(from: lastScalarIdx, to: idx))
+      expectEqual(
+        str.utf16[..<idx].count, str.utf16.distance(from: start, to: idx))
+      expectEqual(
+        str.utf16[idx...].count, str.utf16.distance(from: idx, to: end))
     }
   }
+
+  let nsstring: NSString = "aодиde\u{301}日🧟‍♀️"
+  doIt(nsstring as String)
+
+  let string = "aодиde\u{301}日🧟‍♀️"
+  doIt(string)
 }
+
 #endif // _runtime(_ObjC)
 
+StringIndexTests.test("Exhaustive Index Interchange") {
+  // Exhaustively test aspects of string index interchange
+  func testInterchange(
+    _ str: String,
+    stackTrace: SourceLocStack = SourceLocStack(),
+    showFrame: Bool = true,
+    file: String = #file,
+    line: UInt = #line
+  ) {
+    guard #available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) else {
```

Comment:
Can this be macOS 10.15, etc now?

Reply:
I'm matching what's currently there. I had also tried to update all of String's versions in another commit, but hit a snag during testing (don't recall what it was, probably I wasn't running the beta). I want to get this in for 5.1 in time, so I just followed what we already have throughout the source.